crates/percy-dom/src/patch/apply_patches.rs

```diff
@@ -241,7 +241,13 @@ fn apply_element_patch(
                         }
                     }
                     AttributeValue::Bool(val_bool) => {
-                        if *val_bool {
+                        // Use `set_checked` instead of `{set,remove}_attribute` for the `checked` attribute.
```

Comment:
One use case for the attribute is for anyone that is server-side rendering their site.

So, if you're using `percy-dom` strictly for server-side rendering, with no client-side logic, this would be a regression.

Is there any advantage in not setting the property? Or was this just done to remove seemingly unnecessary behavior?

If there is no strong reason, let's continue to set the attribute since it's useful for server-side rendering.

Reply:
This does seem like a serious issue. I was thinking about a case regarding CSR, not SSR.

Solution proposals are at the bottom that achieve the desired affect without making this CSR case a thorn in a developer's side.

Note: actually, there are a lot of considerations in the original issue #205 that don't regard what a good default for SSR would be, especially when it would actually be the opposite of what I was suggesting... I'll leave a comment on #205 to account for this.

---

## Why do it this way

The reason I did this was to avoid overwriting the default checkedness (HTML checked attribute) when it has been explicitly configured by the developer.

Let's say a developer wanted to set the default checkedness ("why?" is answered later). From the addition to the book:

```rust
let vnode = html! {<input type="checkbox">};

let mut events = VirtualEvents::new();
let (input_node, enode) = vnode.create_dom_node(&mut events);
events.set_root(enode);

// Sets the default checkedness to true by setting the `checked` attribute.
let input_elem = input_node.dyn_ref::<HtmlInputElement>().unwrap();
input_elem.set_attribute("checked", "").unwrap();
```

If the user proceeds to update the VDOM with `let vnode = html! {<input type="checkbox"> checked=false};` then presumably the developer didn't want to override the default checkedness that they configured, they just wanted to set the rendered checkbox state to off.

If we reverted the behavior, there would be no way to specify the default checkedness besides resetting it after every `percy` update.

### Why explicitly configure the default checkedness after forcing the checkbox state?

Configuring form reset state. https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/reset

If the user doesn't _always_ specify the checkedness, then they might rely on the fact that resetting a form which checkedness is not explicitly specified will return the checkbox to the default checkedness.
- e.g. Somewhat contrived client-side-rendering example: The application has a map that holds option-value pairs, it starts off empty. The GUI has a form that holds a bunch of options (with default states), if the user interacts with an input, the map is updated to explicitly contain the user-configured value. (Once the user is done with the form, the application sends the form to the server, which only contains the stuff the user explicitly changed (the server knows what the defaults are, so its able to infer in the rest of the form data itself, if needed).) The GUI only overrides the browser checkboxes's checkedness if there's an explicit value in the map for it. Therefore if the user submits the form (resetting the form and clearing the map) then the default checkedness of the checkbox in the form determines the reset behavior.

If we had `html!`'s `checked` set both the checkedness and default checkedness, then the form reset state would always be the current state for checkboxes. This seems odd as it makes resetting the form useless - it's now a no-op. **(Note that this is currently how `value` works in `percy`, which should probably use whatever solution we decide is best for `checked`.)**

---

### Personal thoughts

Regarding client side rendering: I would be surprised if setting `checked` changed the default checkedness.

But we're between a rock and a hard place. A good default for CSR (follow the application state, lock the GUI state to what is explicitly specified using `html!`) isn't a good default for SSR (let the browser do its thing, don't try to lock the GUI state).

---

## What if `checked` did override default checkedness?

Let's assume that a developer might want to ensure the checkbox has a specific default checkedness (again, this is exclusively considering CSR, not SSR).

They would be forced to reset the default checkedness after each DOM update.

This might be very messy because the code that created the `VirtualNode` for a form might be far from where `percy`'s `update` or `render` functions are called. It's only after `update` or `render` is called can the default checkedness be overridden to be what the developer wanted. This would be a maintenance hazard.

---

# Solutions proposals:

1)
Add a fake `override_default_checked={bool}` attribute to `html!` which sets default checkedness explicitly when desired. Otherwise `checked` sets the HTML attribute as well.

This specific name was chosen to help convey that updating the following

`let vnode = html! {<input type="checkbox" checked=false override_default_checked=true>};`

to

`let vnode = html! {<input type="checkbox" checked=false>};`

would result in a checkbox with false checkedness and false default checkedness, as the override is no longer specified. (I just want to avoid `default_checked`, which would make this confusing. Not specifying `default_checked` would cause the default checkedness to get overridden by `checked` which I wouldn't expect.)

The idea here is to have a reasonable default for CSR and SSR (just set both the checkedness and default checkedness), but allow it to be overridden in weird edge-cases where developers want a specific default checkedness (maybe useful for CSR). This also means that specifying default checkedness is no longer a hacky-looking piece of code.

2)
I'm not familiar with SSR, but can we assume that SSR vnodes will only be created, not `diff`+`patch`'d? In which case we can keep the current behavior for CSR applications that use `diff`+`patch`. We just need to ensure that `checked` sets the HTML attribute when `vnode.to_string()` is called, at a minimum (maybe [`create_dom_node`](https://github.com/chinedufn/percy/blob/a79c609a211820d6da993d25280e290831257973/crates/virtual-node/src/lib.rs#L138-L150) too?).

This doesn't conflict with the above method of configuring default checkedness, as the default checkedness can only be overridden in the DOM after the DOM representation is created.

If `create_dom_node` also exhibited this behavior: unless overridden, the default checkedness will be the first `checked` state, which seems suitable to most scenarios, both SSR and CSR, and can be overridden as previously described at the top. This feels like a sensible default, and it's how it usually works: when serving a normal HTML wesite, the default checkedness is specified in the HTML, which is both the default checkedness AND the initial value. 

_We could add a `default_checked` attribute now or later to avoid the hacky-looking method for setting default checkedness, but it's not important, as I expect changing the default checkedness from the original checkbox state (implying that the default checkedness matters [rare] AND the default checkedness should not be the initial value [rare])._

---

(2) is not sufficient if a server doing SSR ever uses `diff` and `patch`, because the desired state (current checkbox state according to the DOM) isn't described by the HTML, it's in the DOM, but we're only sending the HTML.

I'm leaning towards solution (2) if that's not an issue. (2) feels minimal, intuitive, like the "correct default" (with an escape hatch).

Otherwise, I can't think of anything better than (1) at the moment.

Reply:
> If we reverted the behavior, there would be no way to specify the default checkedness besides resetting it after every percy update.

Percy only updates the element if it has changed.

So, if your element always looks like this:

```rust
html!{ <input type="checked" checked=true /> }
```

then it will always be checked.

So, there is a way to specify the default checkedness. Just set the attribute and never change it. `percy` will never touch it again if the `VirtualNode` never changes.

Not saying that relying on this is intuitive. Just mentioning so that we're designing on top of truthful ground.

---

> The application has a map that holds option-value pairs, it starts off empty. The GUI has a form that holds a bunch of options (with default states)

If you are running `client-side` code you should be expected to render declaratively.

Your state should mimic what you see.

If you want to deviate from that model then don't create the `<input>` element using `percy`.

Just create it yourself and append it to your `percy` element.

```rust
fn render() -> VirtualNode {
    html! { <div id='container-for-non-percy-elements'></div> }
}

fn attach_form() {
    let form = document.create_element("form");
    let input = document.create_element("input");
    // ...

    let container = document.get_element("container-for-non-percy-elements");
    container.append_child(form);
}
```

What I'm proposing above is:

- if you are running `percy` client side, we expect that you are using `percy`'s `html!` macro / `VirtualNode::*` APIs declaratively.
  - As in, whatever your `html!` says is what the user will see

- if you want to deviate from that model, do some outside of `percy`. Create the elements yourself, then append them yourself to a `percy` element.
  - So, say you're at a company that has been using a `jQuery` form library for 20 years. Cool, no problem. Just call a function to append that to your `percy` element
    - `percy` won't touch those elements. it doesn't know about them. it only knows about it's `VirtualNode`s
      - unless of course you delete the `<div id='container-for-non-percy-elements'></div>` that you created using `percy`
      - so, in short, you can embed whatever you want inside of your `percy`-managed elements. If you need to do weird stuff, embed the elements yourself.
        - then it just becomes our job to provide examples and guides that arm people with the knowledge of when and how to leave `percy` to do edge case things. Using regular old HTML primitives. 
          - because people will generally think that they have to use the library for everything. You don't. You can side-step it whenever you want. So, for extreme edge cases like this, we shouldn't make `percy` account for it. The user can just side-step `percy`.

> A good default for CSR (follow the application state, lock the GUI state to what is explicitly specified using html!)
> isn't a good default for SSR (let the browser do its thing, don't try to lock the GUI state).

I addressed here why this isn't true for complex applications: 

https://github.com/chinedufn/percy/issues/205

But, in short, for complex applications SSR rendering must also be based on application state (otherwise you're maintaining two very different rendering environments, double the work), its just that the GUI is an HTML string.
But we want that GUI (HTML string) to be rendered in the same way that we render to the DOM.
Anything that makes CSR and SSR different should be very well justified.

> Let's assume that a developer might want to ensure the checkbox has a specific default checkedness (again, this is exclusively considering CSR, not SSR).

Create the element outside of `percy` and then append it to your `percy`-managed DOM tree.

You get exactly what you want, and `percy` doesn't need a complicated API for a vanishingly small edge case.

Another thing to note about the philosophy behind `percy`: we aren't trying to hide the DOM for you.

The fact that we give you a virtual dom is because we don't think there's a better way to manage tremendous amounts of application state.

We'd much prefer if instead of learning about virtual doms you could just learn HTML/real-DOM stuff.
So, when real-DOM stuff handles the job well we can just point users in that direction. If the use cases is common we can abstract it for convenience.

This is why I appreciate the technical details in your docs.
We don't want people to have to learn `percy`. We want them to learn HTML/DOM. Much more useful knowledge.

> can we assume that SSR vnodes will only be created, not diff+patch

yes.

> We just need to ensure that checked sets the HTML attribute when vnode.to_string()

Making SSR and CSR different is undesirable for reasons mentioned above. Unless there's ever a compelling reason.

> Solutions proposals:

Option 4 - `percy` doesn't handle default checkedness at all and if you want on the client-side the book recommends just creating the DOM element yourself using `web-sys`.
 We some day also create an example of this in the `examples/` directory.

This is better for the 99%+ of users that will never care about any of this.

And better for the <1% that do since instead of teaching them a weird `percy`-specific attribute we can instead teach them that "Hey .. by the way you can just jump outside `percy` and do real DOM stuff if you like ... here's how to do it for default checknedess".
They are now armed with knowledge that will be useful for all future edge cases that they run into. "If `percy` isn't designed for my strange use case, I can just write my own code to handle it!"

vs. "reading through docs learning about 20 percy-specific concepts that didn't need to exist."

I operate under the assumption that our users would prefer to do things themselves correctly in a knowledge-transferable way than to have us give them a weird abstraction.
They just need us to point them at how to do it.